demo/date-time-picker-validation-demos.html

```diff
@@ -0,0 +1,51 @@
+<dom-module id="date-time-picker-validation-demos">
+  <template>
+    <style include="vaadin-component-demo-shared-styles">
+      :host {
+        display: block;
+      }
+    </style>
+
+    <h3>Required</h3>
+    <vaadin-demo-snippet id="date-time-picker-required">
+      <template preserve-content>
+        <vaadin-date-time-picker label="Date and time" required error-message="Please choose a value">
+        </vaadin-date-time-picker>
+      </template>
+    </vaadin-demo-snippet>
```

Comment:
1. For some reason the required indicator doesn't seem to be shown at all. This might be a bug in custom-field (if so, then an issue should be opened about it). If it's not simple or not possible to fix in this PR then just open a new issue about it (also in this repo if this repo needs changes).

2. The background of the date and time picker don't become red when the field becomes invalid. This is probably because invalid state is not propagated to date picker and time picker. Probably we should just propagate it one way with a new observer `__invalidChanged(invalid)` (similar to the `__disabledChanged(disabled)` observer). I think it's ok if both of the fields become red when the date-time-picker is invalid. I think this should be fixed in this PR.

3. The field doesn't become invalid if you focus and unfocus it (unlike e.g. vaadin-text-field or vaadin-date-picker). Also just unfocusing either the inner date picker or time picker should probably trigger validation and make the requried field invalid. If this is somehow complex to do I think it's ok to just open a new issue about this and fix it later.

Reply:
1. It's because the selector for the required indicator is specified as follows: `:host([required]:not([has-value])) [part="label"]::after`. In the `vaadin-dtp` `custom-field` by default `has-value` `T`, so the indicator is not visible. Not sure what would be the best approach to fix it. 

2. Creation of the `__invalidChanged` with propagation of the `invalid` state could lead to problems with imperative setting of the `invalid` prop to `true` on `dtp-date-picker` (partly caused by https://github.com/vaadin/vaadin-date-picker/issues/696). As the validation will go deep till the `text-field` and is it doesn't have any validation constraint it will not become `valid` afterwards.
That is only one example of incorrect behaviour. But basically it does mean that we are overriding the validation logic in contained components what is not good because it should be covered in those already.

3. The emptiness of the value on `required` is already under the control of `date-picker` and `time-picker`. I can also create a custom listener for `blur` on the `dtp-custom-field`, but that is going to override the internal components logic, so for now I am just propagating `required` to the fields as it's done on the `custom-field` demos (https://cdn.vaadin.com/vaadin-custom-field/1.0.11/demo/#custom-field-validation-demos). Will lead to the same results, **however the validation of the `dtp` won't run the same time** (I guess it shouldn't as the value wasn't yet changed to any of the internal component, nor for the dtp).

**There is blocker for that: https://github.com/vaadin/vaadin-time-picker/issues/115 as `time-picker` incorrectly operate with `required` and `invalid`.**

Reply:
1. I guess we should add a feature to custom-field so that the `__toggleHasValue(value)` observer (or the relevant parts) can be easily overridden or something like that. Not sure if it's now somehow possible to override the `__toggleHasValue(value)` because it's set in the observ ers getter. It would be good to implement a similar `__toggleHasValue(value)` in DTP and make that set `has-value` both on DTP and on the custom-field. This could be made a new issue though if you can't or don't want to fix it in this PR.

2. Good point. I think we should try to use the existing validation logic in DP and TP but I think we should override/extend their validation checks to also include checking that if the other one (or DTP) is invalid, then the other is also invalid so that always both are invalid or neither is invalid to keep the styling nicely synchronized.

Reply:
3. I think this is ok for now, but should probably open an issue about it since as a user you can notice the inconsistency in behaviour when focusing and unfocusing different kinds of required fields.

Reply:
1. We had a short talk with Tomi about this after the daily and he tested that it actually seems to work if you just override `dtp.$.customField.__toggleHasValue` with dev tools (so should be doable probably in `ready()` of DTP (or maybe on the extended custom-field) though it's a bit bad to override "private" API I think it is ok for now in this case. The method in custom field could be renamed to "protected" with a single underscore but then that would also need a release and it would be maybe a bit overkill for this case.

Reply:
Overriden.

Reply:
1. Required indicator sees to work pretty good now though it is hidden when either date or time picker has a value. Imo it should still be shown even when only one of them doesn't have a value, but I don't know if it matters much. Not a blocker. We can open an issue about it if someone else also things it should be changed.

2. Invalid background color seems to work good now.

3. Invalid state and focus handling seems to work great now.